[MO] cli_parser.py fix to accept scalar value for freezing

pavel-esir opened the pull request:
**Analysis and solution**:  cli_parser check was too rigorous and always was accepting shape argument even if values are frozen with a scalar. Corrected this check only if shape is defined.

Ticket: CVS-74323

Code:
* [x]  Comments
* [x]  Code style (PEP8)
* n/a  Transformation generates reshape-able IR
* n/a  Transformation preserves original framework node names

Validation:
* [x]  Unit tests
* n/a  Framework operation tests
* n/a  Transformation tests
* n/a  Model Optimizer IR Reader check

Documentation:
* n/a  Supported frameworks operations list
* n/a  Guide on how to convert the **public** model
* n/a  User guide update

rkazants commented on tools/mo/unit_tests/mo/utils/cli_parser_test.py:
```diff
@@ -469,6 +469,22 @@ def test_get_shapes_several_inputs_several_shapes2(self):
         for i in placeholder_values_ref.keys():
             npt.assert_array_equal(placeholder_values_res[i], placeholder_values_ref[i])
 
```
please also add a check for the case without type: node_name->False.
And in this case please check data type also

pavel-esir replied:
Partially replied in other comment. Here, if user didn't specified data type we can not check it. And we also here do not have access to the original model data types. Such things will be handles in 
https://github.com/openvinotoolkit/openvino/blob/master/tools/mo/openvino/tools/mo/front/freeze_placeholder_value.py#L48-L57

`node_name->False` - we already have such cases in unit-tests. 
https://github.com/openvinotoolkit/openvino/blob/master/tools/mo/unit_tests/mo/utils/cli_parser_test.py#L858-L859

But there was a bug when such freezing was done without specifying any shape in `--input`. So I added only one additional test case for the scenario when there are no shapes in `--input`

rkazants replied:
I think we need to align it with `freeze_placeholder_with_value` option through which we can specify boolean value for placeholders and its type determined correctly. 
I propose to take a model that requires this case (for example, in order to avoid conditional flows) and use --input option instead of `freeze_placeholder_with_value`

rkazants commented on tools/mo/unit_tests/mo/utils/cli_parser_test.py:
```diff
@@ -469,6 +469,22 @@ def test_get_shapes_several_inputs_several_shapes2(self):
+    def test_get_shapes_and_freezing_with_scalar(self):
+        # shapes and value for freezing specified using --input command line parameter
+        argv_input = "inp1,inp2->157"
+        result_shapes, _ = get_placeholder_shapes(argv_input, None)
```
please check data type also

pavel-esir replied:
At this stage if user didn't specified data type we cannot check it. If user did't specified data type we will derive it during 
https://github.com/openvinotoolkit/openvino/blob/master/tools/mo/openvino/tools/mo/front/freeze_placeholder_value.py#L48-L57

rkazants commented on tools/mo/openvino/tools/mo/utils/cli_parser.py:
```diff
@@ -1169,6 +1169,9 @@ def get_placeholder_shapes(argv_input: str, argv_input_shape: str, argv_batch=No
             if '->' not in inp:
                 continue
             shape = placeholder_shapes[inp.split('->')[0]]
+
```
please update help for `--input` with a new example: `node_name->False`

pavel-esir replied:
Agree, updated.

rkazants reviewed (changes requested):
Also need to add case for boolean type

rkazants commented:
@pavel-esir, did you check that the model gets converted OOB after your fix?

pavel-esir commented:
> @pavel-esir, did you check that the model gets converted OOB after your fix?

Yes, of course i checked, and it's successfully converted. @rkazants addressed your comments, please take a look

rkazants commented on tools/mo/openvino/tools/mo/utils/cli_parser.py:
```diff
@@ -288,10 +288,15 @@ def get_common_cli_parser(parser: argparse.ArgumentParser = None):
-                                   'node `node_name1` with the shape [3 4] as an input node and freeze output port 1 '
+                                   'lists. The data type of input node is specified in braces and '
+                                   'can have one of the values: f64 (float64), f32 (float32), f16 (float16), '
+                                   'i64 (int64), i32 (int32), u8 (uint8), boolean. Data type is optional and '
+                                   'in case if it is not specified then it will be derived from the original model. '
+                                   'Example, to set `input_1` with shape [1 100], `sequence_len` scalar input '
+                                   'with value `150`, and boolean input `is_training` with `False` value use the ' 
+                                   'following format: "input_1[1 10],sequence_len->150,is_training->False"'
+                                   'Another example, use the following format to set input port 0 of the node '
```
Without explicit type specification does the current implementation creates `sequence_len` of float type instead of `int`? If yes, this is a bug in MO because `sequence_len` must be of integer type for sure. The same question for is_training.

pavel-esir replied:
>Without explicit type specification does the current implementation creates sequence_len of float type instead of int? If yes, this is a bug in MO because sequence_len must be of integer type for sure.

It will not create `sequence_len` of float type. It will store as string and then will convert to the `dtype` of the original node:
https://github.com/openvinotoolkit/openvino/blob/master/tools/mo/openvino/tools/mo/front/freeze_placeholder_value.py#L48-L57

>The same question for is_training. I propose to automatically determine its type.
If type is not specified we take it from the original graph. The code above contains all answers.

>The previous behaviour I expect was to specify explicitly its type. Now we can make it optional but we need to do it in intelligent way.
No. Previous behavior was the same: type is optional. I didn't change behavior I only mentioned explicitly this in help.

rkazants commented on docs/MO_DG/prepare_model/convert_model/Converting_Model.md:
```diff
@@ -101,18 +101,24 @@ Framework-agnostic parameters:
-                        4],node_name2:1[2]{i32}->[20 15]".
+  --input INPUT         Quoted list of comma-separated input nodes names with shapes, 
+                        data types, and values for freezing. The shape and value are 
+                        specified as space-separated lists. The    data type of input 
```
please remove extra spaces

pavel-esir replied:
removed extra spaces

rkazants commented on docs/MO_DG/prepare_model/convert_model/Converting_Model.md:
```diff
@@ -101,18 +101,24 @@ Framework-agnostic parameters:
+                        f64 (float64), f32 (float32), f16 (float16), i64 (int64), 
+                        i32 (int32), u8 (uint8), boolean (bool). Data type is optional. 
+                        If it's not specified explicitly then there are two options: 
+                        if input node is a Parameter data type is inherited from the 
```
`,` is missed; `parameter`; `taken`:
if input node is a parameter, data type is taken from the ...

pavel-esir replied:
Agreed and rephrased

rkazants commented on docs/MO_DG/prepare_model/convert_model/Converting_Model.md:
```diff
@@ -101,18 +101,24 @@ Framework-agnostic parameters:
+                        if input node is a Parameter data type is inherited from the 
+                        original node dtype, if input node is not a Parameter data type 
+                        is set to f32. Example, to set `input_1` with shape [1 100], 
+                        and Parameter node `sequence_len` with   scalar input with value `150`, 
```
extra spaces

pavel-esir replied:
Thanks for noticing! Removed

rkazants commented on tools/mo/openvino/tools/mo/front/freeze_placeholder_value.py:
```diff
@@ -48,6 +48,7 @@ def replace_sub_graph(self, graph: Graph, match: dict):
             try:
                 if data_type != np.bool:
                     value = mo_array(string_value, dtype=data_type)
+                # TODO: investigate why bool is allowed only for TF
```
investigate why boolean type is allowed only for TensorFlow only

pavel-esir replied:
done

rkazants commented on tools/mo/unit_tests/mo/utils/cli_parser_test.py:
```diff
@@ -15,7 +15,7 @@
 from openvino.tools.mo.utils.cli_parser import get_placeholder_shapes, get_tuple_values, get_mean_scale_dictionary, \
```
do we need npt import?

pavel-esir replied:
no we don't need it. removed

rkazants reviewed (commented):
Left some comments to fix. The other code looks good
